### PR TITLE
`ty::Unevaluated`: dealing with unused substs

### PR DESCRIPTION
**Pull request by lcnr**
blocked on https://github.com/rust-lang/rust/pull/87280

still not too happy with what I wrote here, pls review :heart:

### CODE REVIEW

**Review by jyn514 — COMMENTED** [review]
This is sort of like lazy normalization, but only for constants, right? What made it easier for constants than for all types?

**Comment by lcnr** [comment]
It's a bit stronger than "full lazy norm", in that you can't look at the generic args of constants at all, but also quite a bit simpler, as it doesn't really interact with our type system.

The biggest reason this is easier is that the affected parts of const generics are still unstable, so we don't have to worry about causing some subtle breakage. @matthewjasper did try to implement lazy norm and afaict found some annoying regressions: https://gist.github.com/matthewjasper/358245b033a1b7bd158acfe0c08d0d09

**Comment by lcnr** [comment]
rust-lang/rust#87280 now landed so this is ready for review/merge.

cc @nikomatsakis

File: src/constants.md [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+
+To deal with this, we intend to look at the generic parameters explicitly mentioned
+by the constants and then search the predicates of its parents to figure out which
+of the other generic parameters are reachable by our constant.
+
+**TODO**: Expand this section once the parameter filtering is implemented.
+
+As constants can be part of their parents `where`-clauses, we mention unevaluated
+constants in their parents predicates. It is therefore necessary to mention unevaluated
+constants without being able to first compute the generic parameters
```
Comment (nikomatsakis) [inline src/constants.md:52]:
```suggestion
constants before we have computed the generic parameters
```

**Review by nikomatsakis — APPROVED** [review]
This seems good. I left one suggestion.

**Comment by lcnr** [comment]
opened #1213

**Pull request merged**